### ISSUE TITLE
Test::More plan() requires explicit `scalar`?  Behaviour difference wrt Test-Simple-1.302136

### ISSUE THREAD

**Issue by cxw42** [perl11/cperl#364]
First-time cperl user, here courtesy of the [cpantesters report](http://www.cpantesters.org/cpan/report/4e8bea0e-61fa-11e8-8fb1-ef5133556b3f) on my first CPAN distribution, [Text::PerlPP](https://metacpan.org/release/CXW/Text-PerlPP-0.500002).  All tests pass on p5p perl 5.10.1 (cygwin x86) and 5.26.1 (cygwin x64) (both perlbrew; CPAN Test::More 1.302136).  However, the `plan` call fails on cperl 5.26.2c (cygwin x64) (perlbrew; Test::More 1.401015c).

My test code has an [array of testcases](https://github.com/interpreters/perlpp/blob/4328bde9d30624323ec096168058302ab8afc859/t/02-basic.t#L10).  I then `plan tests => count_tests(\@testcases, 1, 2);` ([here](https://github.com/interpreters/perlpp/blob/4328bde9d30624323ec096168058302ab8afc859/t/02-basic.t#L42)) using a function to count the tests (some list entries have multiple tests).  On cperl, `plan` gives me the error `Got an undefined number of tests`.

I have the following MCVE.  Save as `foo.t` and run as `perl foo.t` for the failure or `perl foo.t 1` for success.

    use 5.010001; use strict; use warnings; use Test::More;
    
    sub count_tests { return scalar @{+shift}; }
    
    my @testcases=qw(foo);
    
    plan tests => count_tests(\@testcases) unless @ARGV;        # fails
    plan tests => scalar count_tests(\@testcases) if @ARGV;     # succeeds
    
    # Workaround:
    #my $ntests = count_tests(\@testcases);
    #plan tests => $ntests;                          # succeeds
    #plan tests => scalar $ntests;                   # succeeds
    
    ok(!!1, 'yep');

Why do I need to specify `scalar` on the `plan` line, even though `count_tests()` is returning a scalar?

I do not know if this is intended behaviour or a bug.  Obviously, I have a workaround :) , but I would still like to know.  This probably won't be my last CPAN release, and I would like to write cperl-compatible test cases.

I searched the issues and didn't find this one.  I did see discussion of strict typing in #153, and I wonder if that might be similar.  Also, #255 says `TODO Test::Script Have_Plan oddity in Test::Simple`, but doesn't explain what oddity that was.  Was it indeed this one?

Thanks very much!

**Comment by rurban** [comment]
This is a FAQ entry. See https://perl11.org/cperl/STATUS.html under FAQ.
tl;dr: The plan argument is typed, it can only be a number.

`count_tests` returns `scalar` (type `:Number`), but the type-optimizer is/was not so good yet, to infer that up to the caller. I think it is now, but to write compat. code, the plan and skip args need to be typed. type :Any is illegal.
